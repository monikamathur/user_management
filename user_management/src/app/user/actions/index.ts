import * as UserActions from './user.actions';
import * as UsersApiActions from './user-api.actions';
import * as CollectionApiActions from './collection-api.actions';
import * as CollectionPageActions from './collection-page.actions';
import * as FindUserPageActions from './find-user-page.actions';
import * as SelectedUserPageActions from './selected-user-page.actions';
import * as ViewUserPageActions from './view-user-page.actions';

export {
  UserActions,
  UsersApiActions,
  CollectionApiActions,
  CollectionPageActions,
  FindUserPageActions,
  SelectedUserPageActions,
  ViewUserPageActions,
};
