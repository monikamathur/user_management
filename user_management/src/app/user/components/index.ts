export * from './user-detail.component';
export * from './user-preview.component';
export * from './user-preview-list.component';
export * from './user-search.component';
