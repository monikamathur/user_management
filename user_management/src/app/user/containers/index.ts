export * from './collection-page.component';
export * from './find-user-page.component';
export * from './selected-user-page.component';
export * from './view-user-page.component';
