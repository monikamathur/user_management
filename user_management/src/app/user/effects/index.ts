export * from './user.effects';
export * from './collection.effects';
