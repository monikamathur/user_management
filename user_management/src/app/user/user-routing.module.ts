import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import {
  CollectionPageComponent,
  FindUserPageComponent,
  ViewUserPageComponent,
} from './containers';

export const routes: Routes = [
  {
    path: 'find',
    component: FindUserPageComponent,
    data: { title: 'Find user' },
  },
  {
    path: ':id',
    component: ViewUserPageComponent,
    data: { title: 'User details' },
  },
  {
    path: '',
    component: CollectionPageComponent,
    data: { title: 'Collection' },
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class UsersRoutingModule {}
