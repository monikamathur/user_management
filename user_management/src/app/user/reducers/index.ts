import { User } from '../models';
import {
  createSelector,
  createFeatureSelector,
  combineReducers,
  Action,
} from '@ngrx/store';
import * as fromSearch from '../reducers/search.reducer';
import * as fromUsers from '../reducers/user.reducer';
import * as fromCollection from '../reducers/collection.reducer';
import * as fromRoot from '../reducers';

export const usersFeatureKey = 'user';

export interface UsersState {
  [fromSearch.searchFeatureKey]: fromSearch.State;
  [fromUsers.usersFeatureKey]: fromUsers.State;
  [fromCollection.collectionFeatureKey]: fromCollection.State;
}

export interface State  {
  [usersFeatureKey]: UsersState;
}

/** Provide reducer in AoT-compilation happy way */
export function reducers(state: UsersState | undefined, action: Action) {
  return combineReducers({
    [fromSearch.searchFeatureKey]: fromSearch.reducer,
    [fromUsers.usersFeatureKey]: fromUsers.reducer,
    [fromCollection.collectionFeatureKey]: fromCollection.reducer,
  })(state, action);
}

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `user` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 *   constructor(state$: Observable<State>) {
 *     this.usersState$ = state$.pipe(select(getUsersState));
 *   }
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const selectUsersState =
  createFeatureSelector<UsersState>(usersFeatureKey);

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const selectUserEntitiesState = createSelector(
  selectUsersState,
  (state) => state.user
);

export const selectSelectedUserId = createSelector(
  selectUserEntitiesState,
  fromUsers.selectId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reduces boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: selectUserIds,
  selectEntities: selectUserEntities,
  selectAll: selectAllUsers,
  selectTotal: selectTotalUsers,
} = fromUsers.adapter.getSelectors(selectUserEntitiesState);

export const selectSelectedUser = createSelector(
  selectUserEntities,
  selectSelectedUserId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the user selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const selectSearchState = createSelector(
  selectUsersState,
  (state) => state.search
);

export const selectSearchUserIds = createSelector(
  selectSearchState,
  fromSearch.getIds
);
export const selectSearchQuery = createSelector(
  selectSearchState,
  fromSearch.getQuery
);
export const selectSearchLoading = createSelector(
  selectSearchState,
  fromSearch.getLoading
);
export const selectSearchError = createSelector(
  selectSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of user in the store.
 */
export const selectSearchResults = createSelector(
  selectUserEntities,
  selectSearchUserIds,
  (user, searchIds) => {
    return searchIds
      .map((id) => user[id])
      .filter((user): user is User => user != null);
  }
);

export const selectCollectionState = createSelector(
  selectUsersState,
  (state) => state.collection
);

export const selectCollectionLoaded = createSelector(
  selectCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  selectCollectionState,
  fromCollection.getLoading
);
export const selectCollectionUserIds = createSelector(
  selectCollectionState,
  fromCollection.getIds
);

export const selectUserCollection = createSelector(
  selectUserEntities,
  selectCollectionUserIds,
  (entities, ids) => {
    return ids
      .map((id) => entities[id])
      .filter((user): user is User => user != null);
  }
);

export const isSelectedUserInCollection = createSelector(
  selectCollectionUserIds,
  selectSelectedUserId,
  (ids, selected) => {
    return !!selected && ids.indexOf(selected) > -1;
  }
);
