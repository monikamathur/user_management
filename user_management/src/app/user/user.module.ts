import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { UsersRoutingModule } from './user-routing.module';
import {
  UserDetailComponent,
  UserPreviewComponent,
  UserPreviewListComponent,
  UserSearchComponent,
} from './components';
import {
  CollectionPageComponent,
  FindUserPageComponent,
  SelectedUserPageComponent,
  ViewUserPageComponent,
} from './containers';
import { UserEffects, CollectionEffects } from './effects';

import * as fromUsers from './reducers';
import { AngularMaterialModule } from '../angular-material.module';

export const COMPONENTS = [
  UserDetailComponent,
  UserPreviewComponent,
  UserPreviewListComponent,
  UserSearchComponent,
];

export const CONTAINERS = [
  FindUserPageComponent,
  ViewUserPageComponent,
  SelectedUserPageComponent,
  CollectionPageComponent,
];

@NgModule({
  imports: [
    CommonModule,
    AngularMaterialModule,
    UsersRoutingModule,

    /**
     * StoreModule.forFeature is used for composing state
     * from feature modules. These modules can be loaded
     * eagerly or lazily and will be dynamically added to
     * the existing state.
     */
    StoreModule.forFeature(fromUsers.usersFeatureKey, fromUsers.reducers),

    /**
     * Effects.forFeature is used to register effects
     * from feature modules. Effects can be loaded
     * eagerly or lazily and will be started immediately.
     *
     * All Effects will only be instantiated once regardless of
     * whether they are registered once or multiple times.
     */
    EffectsModule.forFeature([UserEffects, CollectionEffects]),
  ],
  declarations: [COMPONENTS, CONTAINERS],
})
export class UsersModule {}
