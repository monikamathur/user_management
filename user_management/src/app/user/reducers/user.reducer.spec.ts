import { reducer } from '../reducers/user.reducer';
import * as fromUsers from '../reducers/user.reducer';
import {
  UsersApiActions,
  UserActions,
  ViewUserPageActions,
  CollectionApiActions,
} from '../actions';
import { User, generateMockUser } from '../models';

describe('UsersReducer', () => {
  const user1 = generateMockUser();
  const user2 = { ...user1, id: '222' };
  const user3 = { ...user1, id: '333' };
  const initialState: fromUsers.State = {
    ids: [user1.id, user2.id],
    entities: {
      [user1.id]: user1,
      [user2.id]: user2,
    },
    selectedUserId: null,
  };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      // expect(result).toMatchSnapshot();
    });
  });

  describe('SEARCH_COMPLETE & LOAD_SUCCESS', () => {
    type UsersActions =
      | typeof UsersApiActions.searchSuccess
      | typeof CollectionApiActions.loadUsersSuccess;
    function noExistingUsers(
      action: UsersActions,
      usersInitialState: any,
      user: User[]
    ) {
      const createAction = action({ user });

      const result = reducer(usersInitialState, createAction);

      // expect(result).toMatchSnapshot();
    }

    function existingUsers(
      action: UsersActions,
      usersInitialState: any,
      user: User[]
    ) {
      // should not replace existing user
      const differentUser2 = { ...user[0], foo: 'bar' };
      const createAction = action({ user: [user[1], differentUser2] });

      const result = reducer(usersInitialState, createAction);

      // expect(result).toMatchSnapshot();
    }

    it('should add all user in the payload when none exist', () => {
      noExistingUsers(UsersApiActions.searchSuccess, initialState, [
        user1,
        user2,
      ]);

      noExistingUsers(CollectionApiActions.loadUsersSuccess, initialState, [
        user1,
        user2,
      ]);
    });

    it('should add only user when user already exist', () => {
      existingUsers(UsersApiActions.searchSuccess, initialState, [
        user2,
        user3,
      ]);

      existingUsers(CollectionApiActions.loadUsersSuccess, initialState, [
        user2,
        user3,
      ]);
    });
  });

  describe('LOAD', () => {
    const expectedResult = {
      ids: [user1.id],
      entities: {
        [user1.id]: user1,
      },
      selectedUserId: null,
    };

    it('should add a single user, if the user does not exist', () => {
      const action = UserActions.loadUser({ user: user1 });

      const result = reducer(fromUsers.initialState, action);

      // expect(result).toMatchSnapshot();
    });

    it('should return the existing state if the user exists', () => {
      const action = UserActions.loadUser({ user: user1 });

      const result = reducer(expectedResult, action);

      // expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    it('should set the selected user id on the state', () => {
      const action = ViewUserPageActions.selectUser({ id: user1.id });

      const result = reducer(initialState, action);

      // expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('selectId', () => {
      it('should return the selected id', () => {
        const result = fromUsers.selectId({
          ...initialState,
          selectedUserId: user1.id,
        });

        // expect(result).toMatchSnapshot();
      });
    });
  });
});
