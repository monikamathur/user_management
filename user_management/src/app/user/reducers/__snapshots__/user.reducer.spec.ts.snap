// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersReducer LOAD should add a single user, if the user does not exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer LOAD should return the existing state if the user exists 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all user in the payload when none exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all user in the payload when none exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only user when user already exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only user when user already exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedUserId": null,
}
`;

exports[`UsersReducer SELECT should set the selected user id on the state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedUserId": "1",
}
`;

exports[`UsersReducer Selectors selectId should return the selected id 1`] = `"1"`;

exports[`UsersReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "selectedUserId": null,
}
`;
